File: cacms/models/patient.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, List, Optional
from sqlalchemy import Text, Boolean, Integer, CheckConstraint, UniqueConstraint, ForeignKey, text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID
from cacms.database import Base

if TYPE_CHECKING:
    from cacms.models.appointment import Appointment


class Patient(Base):
    __tablename__ = "patients"
    __table_args__ = (
        UniqueConstraint("clinic_id", "phone", name="uq_patients_clinic_phone"),
        CheckConstraint("gender IN ('male', 'female', 'other')", name="ck_patients_gender"),
    )

    patient_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    name: Mapped[str] = mapped_column(Text, nullable=False)
    phone: Mapped[str] = mapped_column(Text, nullable=False, index=True)
    age: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    gender: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    address: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    consent_given: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    consent_date: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    clinic_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("clinics.clinic_id"), nullable=False, index=True
    )
    created_at: Mapped[datetime] = mapped_column(nullable=False, server_default=text("now()"))

    appointments: Mapped[List["Appointment"]] = relationship("Appointment", back_populates="patient")

File: cacms_flutter/lib/core/models/patient.dart
/// Matches backend PatientOut schema.
class Patient {
  const Patient({
    required this.patientId,
    required this.name,
    required this.phone,
    this.age,
    this.gender,
    this.address,
    required this.consentGiven,
    this.consentDate,
    required this.createdAt,
  });

  final String patientId;
  final String name;
  final String phone;
  final int? age;
  final String? gender;
  final String? address;
  final bool consentGiven;
  final DateTime? consentDate;
  final DateTime createdAt;

  factory Patient.fromJson(Map<String, dynamic> json) => Patient(
        patientId: json['patient_id'] as String,
        name: json['name'] as String,
        phone: json['phone'] as String,
        age: json['age'] as int?,
        gender: json['gender'] as String?,
        address: json['address'] as String?,
        consentGiven: json['consent_given'] as bool,
        consentDate: json['consent_date'] != null
            ? DateTime.parse(json['consent_date'] as String)
            : null,
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  Map<String, dynamic> toJson() => {
        'patient_id': patientId,
        'name': name,
        'phone': phone,
        'age': age,
        'gender': gender,
        'address': address,
        'consent_given': consentGiven,
        'consent_date': consentDate?.toIso8601String(),
        'created_at': createdAt.toIso8601String(),
      };
}
